File: Algorithms/Medium/GroupAnagrams.go
package medium

import (
	"sort"
	"strings"
)

// Time complexity: O(nklog(k))
//   k - length of string
//   n - length of array
// Space complexity: O(n)
func groupAnagrams(strs []string) [][]string {
	result := make([][]string, 0)
	hashMap := make(map[string][]string)

	for _, str := range strs {
		sorted := sortWord(str)
		hashMap[sorted] = append(hashMap[sorted], str)
	}

	for _, v := range hashMap {
		result = append(result, v)
	}

	return result
}

func sortWord(word string) string {
	sorted := strings.Split(word, "")
	sort.Strings(sorted)
	return strings.Join(sorted, "")
}

File: Algorithms/Medium/49.Group-Anagrams/GroupAnagrams.py
# Time complexity: O(mnlogn)
#   n - length of string
#   m - length of array
# Space complexity: O(n)
class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        hash_map = {}

        for s in strs:
            s_sorted = "".join(sorted(s))
            if s_sorted not in hash_map:
                hash_map[s_sorted] = []

            hash_map[s_sorted].append(s)

        return list(hash_map.values())